presto-main/src/main/java/com/facebook/presto/operator/RemoteProjectOperator.java

```diff
@@ -0,0 +1,196 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.metadata.FunctionManager;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.InputReferenceExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class RemoteProjectOperator
+        implements Operator
+{
+    private final OperatorContext operatorContext;
+    private final FunctionManager functionManager;
+    private final List<RowExpression> projections;
+
+    private final CompletableFuture<Block>[] result;
+
+    private boolean finishing;
+
+    private RemoteProjectOperator(OperatorContext operatorContext, FunctionManager functionManager, List<RowExpression> projections)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.functionManager = requireNonNull(functionManager, "functionManager is null");
+        this.projections = ImmutableList.copyOf(requireNonNull(projections, "projections is null"));
+        this.result = new CompletableFuture[projections.size()];
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && !processingPage();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        checkState(!processingPage(), "Still processing previous input");
+        requireNonNull(page, "page is null");
+        for (int channel = 0; channel < projections.size(); channel++) {
+            RowExpression projection = projections.get(channel);
+            Type type = projection.getType();
+            if (projection instanceof InputReferenceExpression) {
+                result[channel] = completedFuture(page.getBlock(((InputReferenceExpression) projection).getField()));
+            }
+            else if (projection instanceof CallExpression) {
+                CallExpression remoteCall = (CallExpression) projection;
```

Comment:
does it make sense to check it's a remote function call ?

Reply:
If this is not a remote call the executor will throw exception anyways. Checking whether it's remote here means we need to fetch the function metadata. The overhead is not necessary.